File: modules/kairos/android/src/main/java/studio/zojer/kairos/engine/KairosModule.kt
package studio.zojer.kairos.engine

import android.content.Context
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.security.MessageDigest
import org.json.JSONObject

// Access is serialized across module instances, including fast refresh.
internal object KairosNative {
  private var initialized = false
  private var loaded = false
  external fun initialize(path: String, atlasPath: String): String
  external fun search(query: String): String
  external fun calculate(request: String): String

  private fun initializeIfNeeded(context: Context) {
    if (!loaded) {
      System.loadLibrary("kairos_jni")
      loaded = true
    }
    if (!initialized) {
      val manifest = JSONObject(context.assets.open("Ephemeris/manifest.json").bufferedReader().use { it.readText() })
      val directory = File(context.filesDir, "kairos-ephemeris")
      check(directory.isDirectory || directory.mkdirs()) { "Cannot create ephemeris directory" }
      for (name in manifest.keys()) {
        val bytes = context.assets.open("Ephemeris/$name").use { it.readBytes() }
        val hash = MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }
        check(hash == manifest.getString(name)) { "Ephemeris checksum failed: $name" }
        // Re-extract on process startup; no stale file survives an app update.
        File(directory, name).writeBytes(bytes)
      }
      val atlas = File(context.filesDir, "kairos-atlas.db")
      context.assets.open("Atlas/atlas.db").use { input -> atlas.outputStream().use { input.copyTo(it) } }
      initialize(directory.absolutePath, "sqlite://${atlas.absolutePath}?mode=ro")
      initialized = true
    }
  }

  @Synchronized
  fun chart(context: Context, request: String): String {
    initializeIfNeeded(context)
    return calculate(request)
  }

  @Synchronized
  fun locations(context: Context, query: String): String {
    initializeIfNeeded(context)
    return search(query)
  }
}

class KairosModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("Kairos")
    AsyncFunction("searchLocations") { query: String ->
      val context = appContext.reactContext ?: error("React context unavailable")
      KairosNative.locations(context, query)
    }
    AsyncFunction("calculateChart") { request: String ->
      val context = appContext.reactContext ?: error("React context unavailable")
      KairosNative.chart(context, request)
    }
  }
}

File: modules/kairos/ios/KairosModule.swift
import ExpoModulesCore
import CryptoKit

public class KairosModule: Module {
    // Process lifetime: fast refresh must not reinitialize the Rust globals.
    private static let engineQueue = DispatchQueue(label: "studio.zojer.kairos.engine")
    private static var initialized = false

    public func definition() -> ModuleDefinition {
        Name("Kairos")
        AsyncFunction("calculateChart") { (request: String) -> String in
            try Self.initializeIfNeeded()
            return try Self.unwrap(request.withCString { kairos_calculate_chart($0) })
        }.runOnQueue(Self.engineQueue)
        AsyncFunction("searchLocations") { (query: String) -> String in
            try Self.initializeIfNeeded()
            return try Self.unwrap(query.withCString { kairos_search_locations($0, 30) })
        }.runOnQueue(Self.engineQueue)
    }

    private static func initializeIfNeeded() throws {
        guard !initialized else { return }
        let moduleBundle = Bundle(for: KairosModule.self)
        guard let url = moduleBundle.url(forResource: "KairosEphemeris", withExtension: "bundle")
                ?? Bundle.main.url(forResource: "KairosEphemeris", withExtension: "bundle"),
              let bundle = Bundle(url: url),
              let manifestURL = bundle.url(forResource: "manifest", withExtension: "json"),
              let manifest = try JSONSerialization.jsonObject(with: Data(contentsOf: manifestURL)) as? [String: String]
        else { throw KairosException("Bundled ephemeris is missing. Rebuild the app.") }
        for (name, expected) in manifest {
            let data = try Data(contentsOf: url.appendingPathComponent(name))
            let actual = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
            guard actual == expected else { throw KairosException("Ephemeris checksum failed: \(name)") }
        }
        guard let atlasURL = bundle.url(forResource: "atlas", withExtension: "db")
        else { throw KairosException("Bundled atlas is missing. Rebuild the app.") }
        _ = try unwrap(url.path.withCString { path in
            "sqlite://\(atlasURL.path)?mode=ro".withCString { atlas in kairos_init(atlas, path) }
        })
        initialized = true
    }

    private static func unwrap(_ result: KairosResult) throws -> String {
        defer { kairos_result_free(result) }
        guard result.success, let data = result.data else {
            throw KairosException(result.error.map { String(cString: $0) } ?? "Unknown engine error")
        }
        return String(cString: data)
    }
}

final class KairosException: GenericException<String> {
    override var reason: String { param }
}
